tools/proto_format/proto_sync.py

```diff
@@ -250,17 +267,17 @@ def GitStatus(path):
 
 
 def Sync(api_root, mode, labels, shadow):
-  pkg_deps = []
   with tempfile.TemporaryDirectory() as tmp:
     dst_dir = pathlib.Path(tmp).joinpath("b")
+    paths = []
     for label in labels:
-      pkg_deps += SyncProtoFile(mode, utils.BazelBinPathForOutputArtifact(label, '.active.proto'),
-                                dst_dir)
-      pkg_deps += SyncProtoFile(
-          mode,
+      paths.append(utils.BazelBinPathForOutputArtifact(label, '.active.proto'))
+      paths.append(
           utils.BazelBinPathForOutputArtifact(
               label, '.next_major_version_candidate.envoy_internal.proto'
-              if shadow else '.next_major_version_candidate.proto'), dst_dir)
+              if shadow else '.next_major_version_candidate.proto'))
+    with mp.Pool() as p:
```

Comment:
I had to resort to `multiprocessing` here to get reasonable performance. It isn't much slower than before on my workstation, but it might be slower on less powerful machines. Ultimately, when we move this all to Bazel native, this performance delta won't matter as we won't be rebuilding unchanged files, but that's post 1.14.0.

Reply:
I think this does affect performance on CI though?

Reply:
CI for format in this run was 12m 52s vs. 8m 59s for a recent PR involving proto changes (https://dev.azure.com/cncf/envoy/_build/results?buildId=36485&view=logs&j=9fd74226-df5b-5f29-39da-76fc341cc763).

It's a regression, but arguably a small fraction of total CI time. Not sure if there is a better way of tackling this without Bazelifying stuff, which is a major project that should happen after the v2 freeze. We need to be able to slice-and-dice the descriptors for the hand edited v3 and combine with v2 upgrade generated shadows somehow.. suggestions welcome :)

Reply:
Given issues we have had with non-deterministic behavior with multi-threading are you confident this is thread safe / deterministic?